pkg/kube/dialer.go

```diff
@@ -0,0 +1,336 @@
+		return nil, err
+	}
+
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	conn, err := p.connectionForPod(pod)
+	if err != nil {
+		return nil, err
+	}
+
+	p.requestID++
+
+	headers := http.Header{}
+	headers.Set(corev1.PortHeader, strconv.Itoa(port))
+	headers.Set(corev1.PortForwardRequestIDHeader, strconv.Itoa(p.requestID))
+
+	headers.Set(corev1.StreamType, corev1.StreamTypeError)
+	errorStream, err := conn.CreateStream(headers)
+	if err != nil {
+		return nil, err
+	}
+
+	headers.Set(corev1.StreamType, corev1.StreamTypeData)
+	dataStream, err := conn.CreateStream(headers)
+	if err != nil {
+		return nil, err
+	}
+
+	return wrapConn(network, address, dataStream, errorStream), nil
+}
+
+// DialContext is a simple wrapper around Dial.
+func (p *PodDialer) DialContext(_ context.Context, network string, address string) (net.Conn, error) {
+	return p.Dial(network, address)
+}
+
+// parseDNS attempts to determine the intended pod to target, currently the
+// following formats are supported:
+//  1. [pod-name].[service-name].[namespace-name].svc.[cluster-domain]
+//  2. [pod-name].[service-name].[namespace-name].svc
+//  3. [pod-name].[service-name].[namespace-name]
+//  4. [pod-name].[service-name]
+//
+// If no cluster-domain is supplied, the dialer's configured domain is assumed.
+// If no namespace-name is supplied, the default namespace is assumed.
+// No validation is done to ensure that the pod is actually referenced by the
+// given service or that the DNS record exists in Kubernetes, instead this assumes
+// that things are set up properly in Kubernetes such that the FQDN passed here
+// matches Kubernetes' own DNS records and a pod within the cluster would be able
+// to resolve the same FQDN to the pod that we do.
+//
+// These assumptions allow us to use this dialer to issues requests *as if* we are
+// in the Kubernetes network even from outside of it (i.e. in tests that attempt to
+// connect to a pod at a given DNS address).
+func (p *PodDialer) parseDNS(fqdn string) (pod types.NamespacedName, port int, err error) {
+	addressPort := strings.Split(fqdn, ":")
+	if len(addressPort) != 2 {
+		err = ErrNoPort
+		return
+	}
+
+	port, err = strconv.Atoi(addressPort[1])
+	if err != nil {
+		return
+	}
+
+	fqdn = addressPort[0]
+	fqdn = strings.TrimSuffix(fqdn, "."+p.clusterDomain)
+	fqdn = strings.TrimSuffix(fqdn, "."+svcLabel)
+	labels := strings.Split(fqdn, ".")
+
+	// since we only dial pods we require 2 labels
+	// (assuming we're trying to dial a pod in the
+	// default namespace) or 3 (for a pod outside
+	// of the default namespace)
+	switch len(labels) {
+	case 2:
+		pod.Namespace = defaultNamespace
+	case 3:
+		pod.Namespace = labels[2]
+	default:
+		err = ErrInvalidPodFQDN
+		return
+	}
+
+	// service - labels[1]
+	pod.Name = labels[0]
+	return
+}
+
+// mutex must be held while calling this
+func (p *PodDialer) connectionForPod(pod types.NamespacedName) (httpstream.Connection, error) {
+	if conn, ok := p.connections[pod]; ok {
```

Comment:
I don't see connections ever getting cleaned up from `p.connections`. Rather than worrying about leaking connections and goroutines, I think we'll be better served by eating the overhead of additional connections and having each connection from `connectionForPod` manage it's own lifecycle by closing the parent connection whenever the connection itself is closed. WDYT?

Reply:
I'll see if I can do something a bit trickier, but with a bit less overhead. Going to just add some reference counting in the parent connection. When a stream is closed then it decrements its parent reference count, when it hits 0 it'll get pruned from the map.